pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,219 @@
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+const (
+	// AutoRegistrationGroupAnnotation on a WorkloadEntry stores the associated WorkloadGroup.
+	AutoRegistrationGroupAnnotation = "istio.io/autoRegistrationGroup"
+	// WorkloadControllerAnnotation on a WorkloadEntry should store the current/last pilot instance connected to the workload for XDS.
+	WorkloadControllerAnnotation = "istio.io/workloadController"
+	// ConnectedAtAnnotation on a WorkloadEntry stores the time in nanoseconds when the associated workload connected to a Pilot instance.
+	ConnectedAtAnnotation = "istio.io/connectedAt"
+	// DisconnectedAtAnnotation on a WorkloadEntry stores the time in nanoseconds when the associated workload disconnected from a Pilot instance.
+	DisconnectedAtAnnotation = "istio.io/disconnectedAt"
+)
+
+func (sg *InternalGen) RegisterWorkload(proxy *model.Proxy, con *Connection) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+
+	// Try to patch, if it fails then try to create
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		setConnectMeta(&cfg, sg.Server.instanceID, con)
+		return cfg
+	})
+	if err == nil {
+		return
+	}
+	// TODO better errors from Get/Patch. NotFound is fine but others should warn.
+
+	// No WorkloadEntry, create one using fields from the associated WorkloadGroup
+	groupCfg := sg.Store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		adsLog.Warnf("auto registration of %v failed: cannot find WorkloadGroup %s/%s", proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+		return
+	}
+	entry := workloadEntryFromGroup(entryName, proxy, groupCfg)
+	setConnectMeta(entry, sg.Server.instanceID, con)
+	_, err = sg.Store.Create(*entry)
+	if err != nil {
+		adsLog.Errorf("auto registration of %v failed: error creating WorkloadEntry: %v", proxy.ID, err)
+	}
+}
+
+func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+	// unset controller, set disconnect time
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		delete(cfg.Annotations, WorkloadControllerAnnotation)
+		cfg.Annotations[DisconnectedAtAnnotation] = strconv.FormatInt(time.Now().UnixNano(), 10)
+		return cfg
+	})
+	if err != nil {
+		adsLog.Warnf("disconnect: failed patching WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+		return
+	}
+
+	gp := time.After(features.WorkloadEntryCleanupGracePeriod)
+	go func() {
```

Comment:
possibly same comment about many goroutines. Wondering if there is some "delayed queue" implementation already we can use maybe? We can do it as an optimization latter though

Reply:
Couldn't find anything out there so I threw something together.